src/streaming/array.writer.cpp

```diff
@@ -224,44 +225,42 @@ zarr::ArrayWriter::write_frame_to_chunks_(std::span<const std::byte> data)
     const auto chunk_offset =
       static_cast<long long>(dimensions->chunk_internal_offset(frame_id));
 
-    for (auto i = 0; i < n_tiles_y; ++i) {
-        // TODO (aliddell): we can optimize this when tiles_per_frame_x_ is 1
-        for (auto j = 0; j < n_tiles_x; ++j) {
-            const auto c = group_offset + i * n_tiles_x + j;
-            auto chunk_ptr = get_chunk_data_(c) + chunk_offset;
-            const auto chunk_end = chunk_ptr + bytes_per_chunk;
-
-            for (auto k = 0; k < tile_rows; ++k) {
-                const auto frame_row = i * tile_rows + k;
-                if (frame_row < frame_rows) {
-                    const auto frame_col = j * tile_cols;
-
-                    const auto region_width =
-                      std::min(frame_col + tile_cols, frame_cols) - frame_col;
-
-                    const auto region_start = static_cast<long long>(
-                      bytes_per_px * (frame_row * frame_cols + frame_col));
-                    const auto nbytes =
-                      static_cast<long long>(region_width * bytes_per_px);
-                    const auto region_stop = region_start + nbytes;
-                    if (region_stop > data.size()) {
-                        LOG_ERROR("Buffer overflow");
-                        return bytes_written;
-                    }
-
-                    // copy region
-                    if (nbytes > std::distance(chunk_ptr, chunk_end)) {
-                        LOG_ERROR("Buffer overflow");
-                        return bytes_written;
-                    }
-                    std::copy(data.begin() + region_start,
-                              data.begin() + region_stop,
-                              chunk_ptr);
-
-                    bytes_written += (region_stop - region_start);
-                }
-                chunk_ptr += static_cast<long long>(bytes_per_row);
+    const auto* data_ptr = data.data();
+    const auto data_size = data.size();
+
+    size_t bytes_written = 0;
+    const auto n_tiles = n_tiles_x * n_tiles_y;
+
+#pragma omp parallel for reduction(+ : bytes_written)
```

Comment:
Check this with a benchmark, but this will scale better if you make bytes_written an atomic and a plain old parallel for.

Most of the work here is not involved with the reduction, and the way reductions get parallelized might be limiting.

Maybe omp figures this out for you, but it's better not to rely on that.

Reply:
I actually see significant improvement with the reduction.

- atomic: 1.081 GiB/s
- reduction: 1.354 GiB/s

I ran multiple times and saw similar numbers.